ggml-qnn.cpp

```diff
@@ -0,0 +1,3425 @@
+    ggml_backend_qnn_context * ctx = (ggml_backend_qnn_context *) backend->context;
+    QNN_LOG_INFO("idx %d, name:%s", ctx->device, g_qnn_mgr[ctx->device].name);
+
+    qnn_instance * instance = (qnn_instance *)g_qnn_mgr[ctx->device].instance;
+    if (instance != nullptr) {
+        std::map<std::string,
+                 std::tuple<Qnn_GraphHandle_t, Qnn_Tensor_t *, Qnn_Tensor_t *,
+                            Qnn_Tensor_t *>>::iterator graph_it;
+        for (graph_it = instance->_qnn_graph_map.begin();
+             graph_it != instance->_qnn_graph_map.end(); graph_it++) {
+            auto & graph_item   = graph_it->second;
+            Qnn_GraphHandle_t & graph_handle = std::get<0>(graph_item);
+            GGML_UNUSED(graph_handle);
+            QNN_LOG_INFO("graph type:%s", graph_it->first.c_str());
+        }
+        instance->_qnn_graph_map.clear();
+
+        instance->qnn_finalize();
+        delete instance;
+        g_qnn_mgr[ctx->device].instance = nullptr;
+    }
+
+    if (g_qnn_mgr[ctx->device].backend != nullptr) {
+        delete backend;
+        g_qnn_mgr[ctx->device].backend = nullptr;
+    }
+    QNN_LOG_INFO("leave %s", __func__);
+}
+
+GGML_CALL static ggml_backend_buffer_type_t ggml_backend_qnn_get_default_buffer_type(ggml_backend_t backend) {
+    ggml_backend_qnn_context * ctx = (ggml_backend_qnn_context *) backend->context;
+
+    return ggml_backend_qnn_buffer_type(ctx->device);
+}
+
+GGML_CALL static ggml_status ggml_backend_qnn_graph_compute(ggml_backend_t backend, ggml_cgraph * cgraph) {
+    enum ggml_status          result = GGML_STATUS_SUCCESS;
+    ggml_backend_qnn_context * ctx   = (ggml_backend_qnn_context *) backend->context;
+    GGML_UNUSED(ctx);
+
+    ggml_compute_params params = {};
+    params.type                = GGML_TASK_TYPE_COMPUTE;
+    params.ith                 = 0;
+    for (int i = 0; i < cgraph->n_nodes; i++) {
+        ggml_tensor * node = cgraph->nodes[i];
+        if (ggml_is_empty(node) || node->op == GGML_OP_RESHAPE ||
+            node->op == GGML_OP_TRANSPOSE || node->op == GGML_OP_VIEW ||
+            node->op == GGML_OP_PERMUTE || node->op == GGML_OP_NONE) {
+            continue;
+        }
+        bool ok = ggml_qnn_compute_forward(ctx, &params, node);
+        if (!ok) {
+            QNN_LOG_DEBUG("error: op not supported %s (%s)\n", node->name, ggml_op_name(node->op));
+        }
+    }
+
+    return result;
+}
+
+GGML_CALL static bool ggml_backend_qnn_supports_op(ggml_backend_t backend,
+                                                   const ggml_tensor * op) {
+    ggml_backend_qnn_context *ctx = (ggml_backend_qnn_context *) backend->context;
+
+    return (ggml_qnn_can_handle_op(ctx, op, false));
+}
+
+GGML_CALL static bool ggml_backend_qnn_offload_op(ggml_backend_t backend,const ggml_tensor * tensor) {
+    ggml_backend_qnn_context * ctx = (ggml_backend_qnn_context *) backend->context;
+
+    return ggml_qnn_compute_forward(ctx, nullptr, (ggml_tensor *) tensor);
+}
```

Comment:
This function only needs to return `true` or `false`, but it must not execute the operation. The purpose of this function is to determine if an operation should be executed in this backend, even if it would require copying weights to the backend memory. As it is, this will either prevent the backend from working entirely, or it will cause many operations to be run twice.

Reply:
Yeah, actually I've tried to make some improvement regarding those comments on my branch, also created a PR in this fork and ask for review several days before, but looks there's no responding on the original author so far. 
Will spend some time on my forknext few weeks, gonna add more operators then.